plugins/libplugin-pay.c

```diff
@@ -2888,11 +2894,13 @@ static void adaptive_splitter_cb(struct adaptive_split_mod_data *d, struct payme
 					     pconstraints->fee_budget,
 					     a->constraints.fee_budget);
 
-			/* Should not fail, mid is less than 55% of original
-			 * amount. fee_budget_a <= 55% of fee_budget_p (parent
-			 * of the new payments).*/
+			/* Should not fail, fee_budget_a <= fee_budget_p
+			 * (parent of the new payments).*/
 			assert(ok);
 
+			/* Payment `a` is the larger split, thus less likely
+			 * to succeed, so start `a` first so that it gets into
+			 * channels earlier.  */
```

Comment:
I intend to add a path diversity paymod later which will `getroute` once for the shortest route, then generate alternate routes from there, by trial-banning some of the nodes/channels on the original route to generate various routes.  That paymod will specifically delay later payments until it has generated alternate routes, as long as the payment goes to the same destination (which is significant for payees with multiple routehints).

The effect is that later splits will go through longer paths, and thus more likely to get delayed reaching the destination relative to the earlier payments, and since it defers later payments to the same destination (it immediately triggers on the earlier payments as soon as it receives the shortest path, then delays subsequent payments to the same destination until it has acquired alternate, and presumably longer, paths).

You want path diversity anyway, regardless of using a non-equal or equal split, since dumping all your payments on the same path is a waste.  If you presplit into 2 parts and route both parts on the exact same path, like the current paymod system does, you just wasted fees and HTLC capacity.

The rule-of-thumb "earlier started is more likely to get available capacity" does not need to be perfect, just good enough to improve payment success rates.

Or just give up on MPP and use JIT routing, it is not as if we have **any** decent research on MPP, whereas JIT routing gets you 90% of what you need from MPP.

Reply:
> I intend to add a path diversity paymod later which will `getroute` once for the shortest route, then generate alternate routes from there, by trial-banning some of the nodes/channels on the original route to generate various routes. That paymod will specifically delay later payments until it has generated alternate routes, as long as the payment goes to the same destination (which is significant for payees with multiple routehints).

Sounds like you're implementing Yen's k-shortest path algorithm :-)

> The effect is that later splits will go through longer paths, and thus more likely to get delayed reaching the destination relative to the earlier payments, and since it defers later payments to the same destination (it immediately triggers on the earlier payments as soon as it receives the shortest path, then delays subsequent payments to the same destination until it has acquired alternate, and presumably longer, paths).

This is far from a guarantee however, given that we don't select by roundtrip latency, but rather costs that our payment incur. So it can very well be that even though a "longer" path has numerically more hops, it is shorter latency-wise. Using the time to compute longer routes as a delay between parts of the same payment also seems strange, given that  there are several orders of magnitude between network latency and the time to compute routes, so we'd be back at the artificial introduction of delays, adding to the overall time to completion.

> The rule-of-thumb "earlier started is more likely to get available capacity" does not need to be perfect, just good enough to improve payment success rates.

I'm not saying it has to be perfect, just wanted to play devils advocate and highlight this assumption as it doesn't have any guarantees.

> You want path diversity anyway, regardless of using a non-equal or equal split, since dumping all your payments on the same path is a waste. If you presplit into 2 parts and route both parts on the exact same path, like the current paymod system does, you just wasted fees and HTLC capacity.

I think the path-diversity discussion is independent of the splitter discussion, but I totally agree that having more diversity is a good thing here. I have a patch that issues the `getroute` calls in sequential order, allowing a call to update the `channel_hints` before collecting exclusions for the next call, which can help avoid oversubscribing channels and introduce path diversity.